File: src/main/java/com/protryon/jasm/instruction/instructions/Lookupswitch.java

package com.protryon.jasm.instruction.instructions;

import com.protryon.jasm.Constant;
import com.protryon.jasm.Method;
import com.protryon.jasm.instruction.Instruction;
import com.protryon.jasm.instruction.psuedoinstructions.Label;
import com.shapesecurity.functional.F;
import com.shapesecurity.functional.Pair;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.ArrayList;

public class Lookupswitch extends Instruction {

    public Label _default;
    public Pair<Integer, Label>[] pairs;
    
    @Override
    public String name() {
        return "Lookupswitch";
    }

    @Override
    public int opcode() {
        return 171;
    }

    @Override
    public void read(boolean wide, ArrayList<Constant> constants, Method method, F<Integer, Label> labelMaker, int pc, DataInputStream in) throws IOException {
        int currentOffset = pc % 4;
        int toRead = 3 - currentOffset;
        in.read(new byte[toRead]); // padding
        _default = labelMaker.apply(pc + in.readInt());
        int npairs = in.readInt();
        if (npairs < 0) {
            throw new UnsupportedOperationException("invalid lookupswitch");
        }
        pairs = (Pair<Integer, Label>[]) new Pair[npairs];
        for (int i = 0; i < npairs; ++i) {
            pairs[i] = Pair.of(in.readInt(), labelMaker.apply(in.readInt() + pc));
        }
    }

    @Override
    public void write(boolean wide, DataOutputStream out, F<Label, Integer> labelIndexer, F<Constant, Integer> constantIndexer, int pc) throws IOException {
        int currentOffset = pc % 4;
        int toWrite = 3 - currentOffset;
        out.write(new byte[toWrite]);
        out.writeInt(labelIndexer.apply(_default) - pc);
        out.writeInt(pairs.length);
        for (Pair<Integer, Label> pair : pairs) {
            out.writeInt(pair.left);
            out.writeInt(labelIndexer.apply(pair.right) - pc);
        }
    }

    @Override
    public int pushes() {
        return 0;
    }

    @Override
    public int pops() {
        // key
        return 1;
    }

    @Override
    public String toString() {
        return "Lookupswitch";
    }

    @Override
    public Instruction fromString(String str) {
        return null;
    }

    @Override
    public boolean isControl() {
        return true;
    }
}
        

File: src/main/java/com/protryon/jasm/instruction/instructions/Lookupswitch.kt
package com.protryon.jasm.instruction.instructions

import com.protryon.jasm.Constant
import com.protryon.jasm.Method
import com.protryon.jasm.instruction.Instruction
import com.protryon.jasm.instruction.psuedoinstructions.Label
import com.shapesecurity.functional.Pair

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.util.ArrayList

class Lookupswitch : Instruction() {

    var default: Label? = null
    var pairs: Array<Pair<Int, Label>>? = null

    override val isControl: Boolean
        get() = true

    override fun name(): String {
        return "Lookupswitch"
    }

    override fun opcode(): Int {
        return 171
    }

    @Throws(IOException::class)
    override fun read(wide: Boolean, constants: ArrayList<Constant<*>>, method: Method, labelMaker: (Int)->Label, pc: Int, inputStream: DataInputStream) {
        val currentOffset = pc % 4
        val toRead = 3 - currentOffset
        inputStream.read(ByteArray(toRead)) // padding
        default = labelMaker.invoke(pc + inputStream.readInt())
        val npairs = inputStream.readInt()
        if (npairs < 0) {
            throw UnsupportedOperationException("invalid lookupswitch")
        }
        pairs = arrayOfNulls<Pair<*, *>>(npairs) as Array<Pair<Int, Label>>
        for (i in 0 until npairs) {
            pairs!![i] = Pair.of(inputStream.readInt(), labelMaker.invoke(inputStream.readInt() + pc))
        }
    }

    @Throws(IOException::class)
    override fun write(wide: Boolean, out: DataOutputStream, labelIndexer: (Label)->Int, constantIndexer: (Constant<*>) -> Int, pc: Int) {
        val currentOffset = pc % 4
        val toWrite = 3 - currentOffset
        out.write(ByteArray(toWrite))
        out.writeInt(labelIndexer.invoke(default!!) - pc)
        out.writeInt(pairs!!.size)
        for (pair in pairs!!) {
            out.writeInt(pair.left)
            out.writeInt(labelIndexer.invoke(pair.right) - pc)
        }
    }

    override fun pushes(): Int {
        return 0
    }

    override fun pops(): Int {
        // key
        return 1
    }

    override fun toString(): String {
        return "Lookupswitch"
    }

    override fun fromString(str: String): Instruction {
        error("unsupported")
    }
}
        